File: PythonConversion/functionPartition.py
import numpy as np
import importlib
importlib.import_module("functionSelectArc")
from functionSelectArc import selectArc
importlib.import_module("functionArcSplit")
from functionArcSplit import arcSplit
importlib.import_module("functionGbound")
from functionGbound import gx_bound

def Partition(x_now, newCell, k, p, c_L, c_U, M, y,d,edge,origin,destination,Len,A1,A3,df_cell, K_newly_added):
    # global K_bar, K_newly_added, d, df_cell
    # global A1, A2, A3, A4, A5
    print("\tInside Partition Function, cell k = ", k)
    print("\tNewCell = ", newCell)
    # Selecting the arc to split
    arc_split = selectArc(x_now, c_L, c_U, M, y,d,edge,origin,destination,A1)
    print("arc_split ", arc_split)
    label = df_cell.at[k, "PI"]
    
    # Calculate ΔL and ΔU
    ΔL, ΔU = arcSplit(x_now, arc_split, k, c_L, c_U, M, y, label,edge,origin,destination,d,Len,A3)

    # Create info for the new cell K+1
    cL_newCell = np.copy(c_L)
    cL_newCell[arc_split] += ΔL
    cU_k = np.copy(c_U)
    cU_k[arc_split] = cU_k[arc_split] - ΔU
    # print("cU_k = ", cU_k)
    # print("2.2")
    # print(df_cell)
    cL_avg = (c_L + cU_k) / 2
    cU_avg = (cL_newCell + c_U) / 2
    yL, gL, SP_L, label_L, path_L = gx_bound(cL_avg, cL_avg + d * x_now, edge, origin,destination)
    # print("yL = ", yL)
    # print("cL_avg = ", cL_avg)
    # print("cL_avg + d * x_now = ", cL_avg + d * x_now)
    yU, gU, SP_U, label_U, path_U = gx_bound(cU_avg, cU_avg + d * x_now, edge, origin,destination)
    # print("label_L ",label_L)

    current_p = df_cell.loc[k, "PROB"]
    # Add information for the new cell K+1
    # print("Before update")
    # print(df_cell)
    df_cell = df_cell.append({'CELL':newCell, 'Y': yU, 'Y_Lk':yU, 'g':gU, 'h':0, 'gL':0, 'LB':cL_newCell, 'UB':c_U, 'PROB':current_p * (ΔU / M[arc_split]), 'PI':label_U},ignore_index=True)
    # print("2.3")
    # print(df_cell)
    # print("k = ", k)
    # print(df_cell.loc[k-1,:])
    # print("cU_k = ", cU_k)
    # Updating information for the revised cell k
    # print("replace ", df_cell.loc[df_cell.CELL == k,'UB'])
    # print("with ", np.array([cU_k]))
    # print("g ", gL, "; h ", 0, "PROB ", current_p * (ΔL / M[arc_split]))
    # print(    df_cell.loc[df_cell.CELL == k,'UB'])
    # print("Before update df_cell")
    # print(df_cell)
    # print(df_cell.CELL == k)
    # print("/n k = ", k)
    # print("cU_k = ", cU_k)
    # print("df_cell.loc[df_cell.CELL == k,'UB'] ", df_cell.loc[df_cell.CELL == k,'UB'])
    #Don't use .loc -- doesn't update correctly when dealing with array.
    df_cell.at[k,'UB'] = cU_k
    # print("df_cell.loc[df_cell.CELL == k,'UB'][0] ", df_cell.loc[df_cell.CELL == k,'UB'][0])
    df_cell.at[k,'Y'] = yL
    df_cell.at[k,'g'] = gL
    df_cell.at[k,'h'] = 0
    df_cell.at[k,'PROB'] = current_p * (ΔL / M[arc_split])
    df_cell.at[k,'PI'] = label_L
    # print("After update df_cell")
    # print(df_cell)
    K_newly_added.append(newCell)

    ΔL /= 2
    ΔU /= 2

    return ΔL, ΔU, arc_split, yL, yU, gL, gU, SP_L, SP_U,df_cell

File: DeletedFiles/functionPartition.jl
# function Partition(x_now, newCell, k, p, c_L, c_U, M, y, yL)
function Partition(x_now, newCell, k, p, c_L, c_U, M, y)
    global K_bar, K_newly_added, d, df_cell
    global A1, A2, A3, A4, A5
        # #Applying Lemma 2
        # yW, gW, SPW = gx_bound(cW, c_g_W, edge)
        # # println("yW = ", findall(yW.>0.9))
        # M_path = broadcast(abs,y-yW).*M
        # S_k_W = findall((M_path.==maximum(M_path)) .& (M_path.>0))
        # # println("S_k_W = ", S_k_W)
    arc_split = selectArc(x_now, c_L, c_U, M, y) #S_k_W[1]
    # println("\nk=", k," ; ",arc_split)
    # println("c_orig = [", c_L[arc_split], " \t", c_U[arc_split],"]")
    
#     end
    
    label = df_cell[k,:PI]
    #ΔL=M-gap of arc_split for revised k 
    #ΔU=M-gap of arc_split for new cell k+1
    #ΔL+ΔL = original k's M-gap of arc_split 
    ΔL, ΔU = ArcSplit(x_now,arc_split,k,c_L, c_U, M,y, label)
    
    #Create info for new cell |K|+1
    cL_newCell = deepcopy(c_L)
    cL_newCell[arc_split] = cL_newCell[arc_split] + ΔL 
    cU_k = deepcopy(c_U)
    cU_k[arc_split] = cU_k[arc_split] - ΔU 
    # M_ = zeros(Len)
    # println("\t k: ", c_L[arc_split], " \t", cU_k[arc_split])
    # println("\t |K|+1: ", cL_newCell[arc_split], " \t", c_U[arc_split])
    
    
    cL_avg = (c_L+cU_k)/2
    cU_avg = (cL_newCell+c_U)/2
    yL, gL, SP_L, T_L, pred_L, label_L, path_L = gx_bound(cL_avg, cL_avg+d.*x_now, edge)
    yU, gU, SP_U, T_U, pred_U, label_U, path_U = gx_bound(cU_avg, cU_avg+d.*x_now, edge)

    current_p = df_cell[k,:PROB]
    #Add to df info for new cell K+1
    push!(df_cell, (newCell, yU, yU, gU, 0, 0, cL_newCell, c_U, current_p*(ΔU/M[arc_split]), label_U))
    # println(df_cell[newCell,:])
    #Updating Info for revised cell k
    df_cell[k,:UB] = cU_k
    df_cell[k,:Y] = yL
    df_cell[k,:g] = gL
    df_cell[k,:h] = 0
    df_cell[k,:PROB] = current_p*(ΔL/M[arc_split]) #p/2
    df_cell[k,:PI] = label_L
    # println(df_cell[k,:])
    push!(K_newly_added, newCell)

    # break
    # exit()
    ΔL = ΔL/2
    ΔU = ΔU/2
    return ΔL, ΔU, arc_split, yL, yU, gL, gU, SP_L, SP_U
end
